File: examples/rastrigin.rs
//! Rastrigin: minimize a real-valued function with many local minima.
//!
//! Shows a real-valued genome, minimization, parallel evaluation and a time limit. The global
//! minimum is 0, at the origin.
//!
//! Uniform mutation finds the basin of the global minimum, but it's a blunt tool for the last
//! digits: the Gaussian and polynomial mutations planned for 0.2 fine-tune much better.
//!
//! ```text
//! cargo run --release --example rastrigin
//! ```

use genoxide::prelude::*;
use std::f64::consts::TAU;
use std::time::Duration;

const DIMENSIONS: usize = 10;

fn rastrigin(x: &Reals) -> f64 {
    10.0 * x.len() as f64
        + x.iter()
            .map(|xi| xi * xi - 10.0 * (TAU * xi).cos())
            .sum::<f64>()
}

fn main() -> Result<()> {
    let ga = Ga::builder(Real::uniform(DIMENSIONS, -5.12..=5.12)?)
        .population_size(200)
        .select(Tournament::new(4)?)
        .crossover(UniformCrossover::new())
        .mutate(UniformMutation::count(1)?)
        .scheme(Scheme::Generational { elitism: 2 })
        .minimize()
        .seed(3)
        .build()?;

    let outcome = Engine::new(ga, rastrigin)
        // worth it for expensive fitness functions; the results are the same either way
        .parallel(true)
        .stop_when(
            Stop::target(1e-3)
                .or(Stop::stagnation(500))
                .or(Stop::time(Duration::from_secs(30))),
        )
        .on_generation(|snapshot| {
            let progress = snapshot.progress();
            if progress.generation() % 200 == 0 {
                let best = progress.best().unwrap_or(Fitness::invalid());
                println!("{:>6}: {best:.6}", progress.generation());
            }
        })
        .run()?;

    println!(
        "\n{:?} after {} generations ({:?}): {:.6}",
        outcome.stop_reason(),
        outcome.generations(),
        outcome.elapsed(),
        outcome.best_fitness()
    );
    println!("at {:.4?}", &outcome.best_genome()[..]);
    Ok(())
}

File: python/examples/rastrigin.py
"""Rastrigin in 30 dimensions: CMA-ES with IPOP restarts, and L-SHADE, with a generation per call
of a vectorized numpy function."""

import numpy as np

import genoxide as gx


def rastrigin(x):
    """x has a genome per row."""
    return 10 * x.shape[1] + np.sum(x**2 - 10 * np.cos(2 * np.pi * x), axis=1)


genome = gx.Real((-5.12, 5.12), length=30)
budget = 1_000_000
for algorithm in (
    gx.Cmaes(genome, restarts="ipop", objective="minimize", seed=1),
    gx.De(genome, l_shade=budget, objective="minimize", seed=1),
):
    result = algorithm.run(rastrigin, batch=True, target=1e-8, evaluations=budget)
    print(
        f"{type(algorithm).__name__}: {result.best_fitness:.3g} after {result.evaluations} "
        f"evaluations, {result.seconds:.2f} s"
    )
